### DISCUSSION TITLE
Custom server for the API?

### DISCUSSION THREAD

**Discussion by nehalist** [vercel/next.js#12921] [Help]
The docs clearly state

> Before deciding to use a custom server please keep in mind that it should only be used when the integrated router of Next.js can't meet your app requirements. A custom server will remove important performance optimizations, like serverless functions and Automatic Static Optimization.

Is using the custom server for additionally providing an API the right thing to do or would it be better to create the server completely independent from Next.js?
Since lots of the pages I'm going to create will be server-side rendered I thought it might be a good idea to integrate the API directly into the Next.js app, but everytime I read this statement I become unsure.

**Comment by jamesmosier** [comment]
My personal preference is to create a separate API and then use that API using one of the Next [data fetching](https://nextjs.org/docs/basic-features/data-fetching) recommendations. You can still server render no matter where your API is. Or when I'm able, I just integrate my API logic directly into my app using [API routes](https://nextjs.org/docs/api-routes/introduction).

Is there some functionality you need that the API routes do not offer?

**Reply by nehalist** [reply]
I'm honestly not sure if API routes are sufficient for my needs. Not that I'm saying I doubt it, but I just have too little experience with it to give a realistic estimation.

In my case I need a pretty complex API, with JWT login, db connections, sockets, all that kind of fun. Is it possible to do that with API routes? What would be an example of something where API routes wouldn't be the right choice?

**Reply by jensmeindertsma** [reply]
What you want to create is fairly complicated and large. I would not recommend using API routes for this. In my opinion, those are meant mostly for abstracting more complicated data fetching and not for sockets / JWT login. It's certainly possible and there are examples out there of people who have a setup similar to yours using API routes. Because you are saying that your are probably not comfortable enough with API routes I would suggest building it with whatever technology you are most comfortable with ( for example Node with Express.js ). You can host it on Vercel and the API queries will be just as fast as with API routes.

**Reply by nehalist** [reply]
Thanks for your reply, but that still leaves the initial question open: do I use a custom express server for firing up the Next.js app (which would include the entire API), or should I develop the server completely independent from Next.js?

**Reply by jensmeindertsma** [reply]
If you are not going to develop the API with Next.js, you shouldn't modify the Next.js server because this has negative downsides. I recommend keeping both completely separate in a way where you fire up both apps separately and then fetch the API endpoint from your Next.js app.
